### PR TITLE
issue #464

### PR DESCRIPTION
**Pull request by grandizzy**
- transfer first pledged NFTs from borrower to pool so settle logic can be applied
- added tests for kick and save by repay and pledge more collateral
- added tests for kick, partially take and save by repay and pledge more collateral

### CODE REVIEW

**Review by mattcushman — APPROVED** [review]
lgtm

File: src/erc721/ERC721Pool.sol [hunk 1]
```diff
@@ -87,8 +87,11 @@ contract ERC721Pool is IERC721Pool, FlashloanablePool {
         uint256 limitIndex_,
         uint256[] calldata tokenIdsToPledge_
     ) external {
+        // move collateral from sender to pool to be properly accounted if auction settled
+        if (tokenIdsToPledge_.length != 0) _transferFromSenderToPool(borrowerTokenIds[borrowerAddress_], tokenIdsToPledge_);
+
         (
-            bool pledge,
+            ,
```
Comment (MikeHathaway) [inline src/erc721/ERC721Pool.sol]: Can the pledge return arg be removed entirely?

Comment (grandizzy) [reply src/erc721/ERC721Pool.sol]: it is still used by ERC20 pool but can remove it from there too and do a similar check, is this what you're thinking too?

Comment (MikeHathaway) [reply src/erc721/ERC721Pool.sol]: I don't have a super strong opinion, but seems like it'd be slightly cleaner over all and a bit more gas and size efficient which would be worth the style inconsistency

Comment (grandizzy) [reply src/erc721/ERC721Pool.sol]: 👍 yep, will do so

**Review by EdNoepel — APPROVED** [review]
Hesitant to transfer tokens up front rather than after state changes, but I don't see a better way to do it.

File: src/erc721/ERC721Pool.sol (hunk 1 above)
Comment (grandizzy) [reply src/erc721/ERC721Pool.sol]: changed with https://github.com/ajna-finance/contracts/pull/465/commits/2ba8340ecd08894de71e2e5e3b5c2b14cc4cc369 removed both pledge and borrow return param. That was a good call, besides reducing contract size it also revealed a bug with transferring when amount to borrow was 0 (intentional 0 borrow loan call allowed to update borrower's loan state)

**Comment by grandizzy** [comment]
> Hesitant to transfer tokens up front rather than after state changes, but I don't see a better way to do it.

I'll make some tries in a follow up PR by breaking current settle auction in 2 parts, one to  just calculate the floor / new collateral and 2nd for moving tokens around. this way we can keep transfer at the end and it could also help for externalizing borrower actions into its own external library

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Pull request merged**